batchspawner/batchspawner.py

```diff
@@ -775,4 +777,106 @@ def state_gethost(self):
         self.log.error("Spawner unable to match host addr in job {0} with status {1}".format(self.job_id, self.job_status))
         return
 
+
+class OARSpawner(UserEnvMixin, BatchSpawnerBase):
+    batch_script = Unicode("""#!/bin/sh
+#OAR -n jupyterhub-singleuser-{{username}}
+#OAR -l nodes=1{% if nprocs     %}/core={{nprocs}}{% endif %}{% if runtime    %},walltime={{runtime}}{% endif %}
+{% if queue      %}#OAR -q {{queue}}{% endif %}
+{% if memory     %}#OAR -p "memnode>={{memory}}"{% endif %}
+{% if options    %}#OAR {{options}}{% endif %}
+
+{{prologue}}
+{{cmd}}
+{{epilogue}}
+""").tag(config=True)
+
+    # use a script to submit to batch queueing system instead of piping it in
+    script_mode = stat.S_IRUSR | stat.S_IWUSR | stat.S_IRGRP | stat.S_IROTH | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH #755 by default
+
+    script_home = os.path.join(os.environ.get("HOME", "/tmp"), "oar_jobs")
+    if not os.path.exists(script_home):
+        os.makedirs(script_home)
+    # outputs job id string
+    batch_submit_cmd = Unicode('/usr/bin/oarsub -d {homedir} -S ').tag(config=True)
+    # outputs job data XML string
+    batch_query_cmd = Unicode('/usr/bin/oarstat -J -j {job_id}').tag(config=True)
+    batch_cancel_cmd = Unicode('/usr/bin/oardel {job_id}').tag(config=True)
+
+    @gen.coroutine
+    def run_command(self, cmd, input=None, env=None):
+        if input: 
+            # Use script file
+            script = os.path.join(self.script_home, "jupyter_%s.sh"%self.get_req_subvars()["username"])
+            if env:
+                inputlines = input.split(os.linesep)
+                input = ""
+                for idx, line in enumerate(inputlines):
+                    if line.startswith("#"):
+                        input += line + os.linesep
+                    else:
+                        break
+                input += os.linesep.join(["export %s=%s"%(k, v) for k, v in env.items()]) + os.linesep
```

Comment:
This isn't something for you to do, but I'll look into adding a method for the script templates to include the environment variables, so that this can be made part of the template. Don't worry for now though - don't let this stop this PR yet.